subsys/canbus/isotp/isotp.c

```diff
@@ -692,60 +681,35 @@ int isotp_recv_net(struct isotp_recv_ctx *ctx, struct net_buf **buffer,
 	return *(uint32_t *)net_buf_user_data(buf);
 }
 
-static inline void pull_frags(struct k_fifo *fifo, struct net_buf *buf,
-			      size_t len)
-{
-	size_t rem_len = len;
-	struct net_buf *frag = buf;
-
-	/* frags to be removed */
-	while (frag && (frag->len <= rem_len)) {
-		rem_len -= frag->len;
-		frag = frag->frags;
-		k_fifo_get(fifo, K_NO_WAIT);
-	}
-
-	if (frag) {
-		/* Start of frags to be preserved */
-		net_buf_ref(frag);
-		net_buf_pull(frag, rem_len);
-	}
-
-	net_buf_unref(buf);
-}
-
 int isotp_recv(struct isotp_recv_ctx *ctx, uint8_t *data, size_t len,
 	       k_timeout_t timeout)
 {
-	size_t num_copied, frags_len;
-	struct net_buf *buf;
-	int ret;
+	size_t num_copied, rem_len;
+	int err;
 
-	ret = _k_fifo_wait_non_empty(&ctx->fifo, timeout);
-	if (ret) {
-		if (ctx->error_nr) {
-			ret = ctx->error_nr;
+	if (!ctx->recv_buf) {
+		ctx->recv_buf = net_buf_get(&ctx->fifo, timeout);
+		if (!ctx->recv_buf) {
+			err = ctx->error_nr ? ctx->error_nr : ISOTP_RECV_TIMEOUT;
 			ctx->error_nr = 0;
-			return ret;
-		}
 
-		if (ret == -EAGAIN) {
-			return ISOTP_RECV_TIMEOUT;
+			return err;
 		}
-
-		return ISOTP_N_ERROR;
 	}
 
-	buf = k_fifo_peek_head(&ctx->fifo);
+	num_copied = net_buf_linearize(data, len, ctx->recv_buf, 0, len);
 
-	if (!buf) {
-		return ISOTP_N_ERROR;
+	/* find and delete received fragments in the receive net_buf */
+	rem_len = num_copied;
+	while (ctx->recv_buf && (ctx->recv_buf->len <= rem_len)) {
```

Comment:
While this looks like correctly functioning code, you end up traversing the fragment list multiple time since `linearize()` does it too. Why not just skip `linearize()` and instead use `net_buf_pull_mem()` in this loop?

Reply:
Good point. Done.